templates/stamp/utilities.sh

```diff
@@ -942,4 +980,456 @@ set_timezone()
+    if [[ $servers_ok -eq $rows_processed ]] && [[ $master_identified -eq 1 ]];
+    then
+        # valid master
+        echo 1
+    else
+        # not currently master
+        echo 0
+    fi
+}
+
+# check if the target server is a valid master
+is_master_server()
+{
+    encoded_replicated_servers_list=$1  # ip address of the servers participating in the replication
+    local_server_ip=$2                  # get the server ip
+    mysql_admin_user=$3                 # admin mysql user 
+    mysql_admin_user_password=$4        # admin mysql user password
+
+    # initialize other key variables
+    replicated_servers_list=(`echo ${encoded_replicated_servers_list} | base64 --decode`)
+    total_servers=${#replicated_servers_list[@]}
+    server_position=0
+
+    for replicated_server in "${replicated_servers_list[@]}"
+    do
+        if [[ $replicated_server == $local_server_ip ]];
+        then
+            local_server_replicated_position=$server_position
+            break
+        fi
+
+        ((server_position++))
+    done
+
+    if [[ -z $local_server_replicated_position ]];
+    then
+        log "Could not locate the IP address of ${HOSTNAME} (${local_server_ip}) in the configured replication topology ${replicated_servers_list}"
+        exit $REPLICATION_MASTER_MISSING
+    fi
+
+    # identify the positions of the other servers participating in the replication topology
+    second_server_position=`get_next_position $server_position $total_servers`
+    second_server=${replicated_servers_list[$second_server_position]}
+
+    third_server_position=`get_next_position $second_server_position $total_servers`
+    third_server=${replicated_servers_list[$third_server_position]}
+
+    # more defensive: the repladmin will override the file/or create a new file
+    replication_status_file="/tmp/replication_status_$server.csv"
+    if [[ -f $replication_status_file ]];
+    then
+        #log "Cleaning up existing replication file"
+        rm $replication_status_file
+    fi
+
+    # run the repl admin to check the replication status from the perspective of the target server
+    mysqlrpladmin --master=${mysql_admin_user}:${mysql_admin_user_password}@${local_server_ip}:$mysql_server_port --slaves=${mysql_admin_user}:${mysql_admin_user_password}@${second_server}:3306,${mysql_admin_user}:${mysql_admin_user_password}@${third_server}:3306 health --format=csv > $replication_status_file
+
+    # clean up the status file (remove comment lines and warning text)
+    sed -i "/^#/ d" $replication_status_file
+    sed -i "/^WARNING/ d" $replication_status_file
+
+    # assess the replication status
+    is_valid_master=`check_master_status ${local_server_ip} ${replication_status_file}`
+
+    echo $is_valid_master
+}
+
+#############################################################################
+# Mysql Data Directory Move Operation
+#############################################################################
+
+move_mysql_datadirectory()
```

Comment:
(this requires in-person discussions and should be addressed with a separate change. In other words, I won't block the current change on this). utilities.sh is following a really bad software design pattern. it's evolved into a giant file full of largely unrelated "static" functions. it needs to be split up along the following guidelines

1. common goals (like a collection of apt installers, collection of global utility functions like log(), etc) AND/OR

1. execution paths. in other words, if a function in here is called once then it should be moved closer to the caller rather than stored here. this follows a more object-oriented design where complexity lives as close as possible to the technical domain